File: Java/322. Coin Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] combs = new int[amount + 1];
        Arrays.fill(combs, -1);
        combs[0] = 0;        
        for (int i = 1; i < amount + 1; i++) {
            for (int j = 0; j < coins.length; j++) {
                if (coins[j] <= i && combs[i - coins[j]] != -1) {
                    if (combs[i] != -1)
                        combs[i] = Math.min(combs[i - coins[j]] + 1, combs[i]);
                    else
                        combs[i] = combs[i - coins[j]] + 1;
                }
            }
        }        
        return combs[amount];
    }
}

File: Python/322. Coin Change.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        counts = [amount + 1] * (amount + 1)
        counts[0] = 0
        for i in range(1, amount + 1):
            for coin in coins:
                if coin <= i and counts[i - coin] != -1:
                    counts[i] = min(counts[i - coin] + 1, counts[i])
        return -1 if counts[amount] > amount else counts[amount]
